File: src/instruction/task_create.rs
use anchor_client::anchor_lang::{
    solana_program::{
        instruction::{AccountMeta, Instruction},
        pubkey::Pubkey,
        system_program, sysvar,
    },
    InstructionData,
};
use cronos_program::state::InstructionData as CronosInstructionData;
use cronos_program::{pda::PDA, state::Task};

pub fn task_create(
    task_pda: PDA,
    daemon: Pubkey,
    owner: Pubkey,
    ix: Instruction,
    exec_at: i64,
    stop_at: i64,
    recurr: i64,
) -> Instruction {
    Instruction {
        program_id: cronos_program::ID,
        accounts: vec![
            AccountMeta::new_readonly(sysvar::clock::ID, false),
            AccountMeta::new(daemon, false),
            AccountMeta::new(owner, true),
            AccountMeta::new(task_pda.0, false),
            AccountMeta::new_readonly(system_program::ID, false),
        ],
        data: cronos_program::instruction::TaskCreate {
            ix: CronosInstructionData::from(ix),
            exec_at,
            stop_at,
            recurr,
            bump: task_pda.1,
        }
        .data(),
    }
}

File: web/src/instruction/taskCreate.ts
import { BN, Program } from "@project-serum/anchor";
import {
  PublicKey,
  SystemProgram,
  SYSVAR_CLOCK_PUBKEY,
  TransactionInstruction,
} from "@solana/web3.js";
import { buildInstructionData } from "./utils";
import { Cronos } from "../idl";
import { Account } from "../account";

export type TaskCreateArgs = {
  daemon: PublicKey;
  ix: TransactionInstruction;
  execAt: BN;
  stopAt: BN;
  recurr: BN;
};

export class TaskCreate {
  private account: Account;
  private cronos: Program<Cronos>;

  constructor(account: Account, cronos: Program<Cronos>) {
    this.account = account;
    this.cronos = cronos;
  }

  public async taskCreate({
    daemon,
    ix,
    execAt,
    stopAt,
    recurr,
  }: TaskCreateArgs): Promise<TransactionInstruction> {
    const configPDA = await this.account.config.pda();
    const daemonData = await this.account.daemon.data(daemon);
    const taskPDA = await this.account.task.pda(daemon, daemonData.taskCount);
    const instructionData = buildInstructionData(ix);
    return this.cronos.instruction.taskCreate(
      instructionData,
      execAt,
      stopAt,
      recurr,
      taskPDA.bump,
      {
        accounts: {
          clock: SYSVAR_CLOCK_PUBKEY,
          config: configPDA.address,
          daemon: daemon,
          task: taskPDA.address,
          owner: daemonData.owner,
          systemProgram: SystemProgram.programId,
        },
      }
    );
  }
}
